### PR TITLE
status_reports lambda

### PR DESCRIPTION
**Pull request by nl0**
# Description

# TODO

- [x] [Changelog](CHANGELOG.md) entry

### CODE REVIEW

**Comment by nl0** [comment]
@sir-sigurd i'll probably adjust html layout and add some inline styling, but the rest of the lambda code seems complete

File: lambdas/status_reports/src/t4_lambda_status_reports/__init__.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+
+AWS_REGION = os.getenv("AWS_REGION")
+assert AWS_REGION
+
+
+session = aiobotocore.session.get_session()
+
+
+def create_cfn():
+    return session.create_client("cloudformation", region_name=AWS_REGION)
```
Comment (sir-sigurd) [inline lambdas/status_reports/src/t4_lambda_status_reports/__init__.py:22]: I think it gets region from `AWS_DEFAULT_REGION` env var set by lambda runtime, so it doesn't make sense to pass region explicitly.

File: lambdas/status_reports/src/t4_lambda_status_reports/__init__.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+async def drain(syn, method: str, key: str, names: T.List[str]) -> T.List[dict]:
+    chunks = [
+        names[i:i + CANARIES_PER_REQUEST]
+        for i in range(0, len(names), CANARIES_PER_REQUEST)
+    ]
+    pages = await asyncio.gather(*[
+        getattr(syn, method)(Names=chunk)
+        for chunk in chunks
+    ])
+    return list(itertools.chain(*[p[key] for p in pages]))
```
Comment (sir-sigurd) [inline lambdas/status_reports/src/t4_lambda_status_reports/__init__.py:47]: Consider this:
```suggestion
    return list(itertools.chain.from_iterable(p[key] for p in pages))
```

File: lambdas/status_reports/src/t4_lambda_status_reports/__init__.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+            c.update(ok=state == "PASSED", lastRun=prev["Timeline"]["Completed"])
+
+    return canaries
+
+
+async def get_resources(cfn, stack_name: str) -> T.List[dict]:
+    result = []
+    async for page in cfn.get_paginator("list_stack_resources").paginate(StackName=stack_name):
+        for r in page["StackResourceSummaries"]:
+            result.append(r)
```
Comment (sir-sigurd) [inline lambdas/status_reports/src/t4_lambda_status_reports/__init__.py:108]:
```suggestion
        result.extend(page["StackResourceSummaries"])
```

File: lambdas/status_reports/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+Jinja2==3.1.2
+aiobotocore==2.3.4
+botocore==1.24.21
```
Comment (sir-sigurd) [inline lambdas/status_reports/requirements.txt:3]: I'd put these ones to setup.py as dependencies and generate requirements.txt with [pip-compile](https://pip-tools.readthedocs.io/en/latest/).

Comment (nl0) [reply lambdas/status_reports/requirements.txt:3]: what for?

Comment (sir-sigurd) [reply lambdas/status_reports/requirements.txt:3]: To have indirect dependencies pinned.

Comment (nl0) [reply lambdas/status_reports/requirements.txt:3]: ok i'll have a look

Comment (nl0) [reply lambdas/status_reports/requirements.txt]: done (tho i used py3.8-based venv to compile requirements, bc it's a PITA to install 3.7 on my machine -- it requires x64 arch, i hope it'll work)

Comment (sir-sigurd) [reply lambdas/status_reports/requirements.txt]: Well, it seems to work, but I think it really should be generated with the same Python version (and actually the same arch). 
It doesn't look like installing Python is really hard (https://diewland.medium.com/how-to-install-python-3-7-on-macbook-m1-87c5b0fcb3b5) or maybe you could you Docker. If you don't want to spend time on it, I could regenerate it by myself.

Comment (nl0) [reply lambdas/status_reports/requirements.txt]: > If you don't want to spend time on it, I could regenerate it by myself.

well, yeah, you can, but going forward i think it's time we move to a newer runtime (3.9 or even 3.10 would be nice)

File: lambdas/status_reports/setup.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+setup(
+    name="t4_lambda_status_reports",
+    version="0.0.1",
+    packages=find_packages(where="src"),
+    package_dir={"": "src"},
+    install_requires=[
+        "Jinja2 ==3.1.2",
+        "aiobotocore ==2.3.4",
+        "botocore ==1.24.21",
```
Comment (sir-sigurd) [inline lambdas/status_reports/setup.py:11]: I think either of these are far more common:
```suggestion
        "Jinja2==3.1.2",
        "aiobotocore==2.3.4",
        "botocore==1.24.21",
```
```suggestion
        "Jinja2 == 3.1.2",
        "aiobotocore == 2.3.4",
        "botocore == 1.24.21",
```

File: lambdas/status_reports/src/t4_lambda_status_reports/__init__.py [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+def async_handler(f):
+    @functools.wraps(f)
+    def wrapper(event, context):
+        return asyncio.run(f(event, context))
+
+    return wrapper
+
+
+@async_handler
+async def lambda_handler(*_):
```
Comment (sir-sigurd) [inline lambdas/status_reports/src/t4_lambda_status_reports/__init__.py:268]: I'd use this for uniformity, but that's up to you.

```suggestion
async def lambda_handler(event, context):
```

**Review by sir-sigurd — COMMENTED** [review]
What about tests? :slightly_smiling_face:

File: lambdas/status_reports/requirements.txt (hunk 4 above)
Comment (sir-sigurd) [reply lambdas/status_reports/requirements.txt]: Well I think we already discussed that we use 3.7 so we can easily move code between components. 3.10 is not even available as lambda runtime.

File: lambdas/status_reports/src/t4_lambda_status_reports/__init__.py [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+    return result
+
+
+async def get_stack_data(cfn, stack_name: str) -> dict:
+    resp = await cfn.describe_stacks(StackName=stack_name)
+    return resp["Stacks"][0]
+
+
+jenv = jinja2.Environment(autoescape=jinja2.select_autoescape())
+# TODO: styling
```
Comment (sir-sigurd) [inline lambdas/status_reports/src/t4_lambda_status_reports/__init__.py:116]: What about styling?

File: lambdas/status_reports/setup.py (hunk 5 above)
Comment (nl0) [reply lambdas/status_reports/setup.py:11]: ok, i just saw this style on some docs page and considered it "official" 🤷

File: lambdas/status_reports/src/t4_lambda_status_reports/__init__.py (hunk 7 above)
Comment (nl0) [reply lambdas/status_reports/src/t4_lambda_status_reports/__init__.py:116]: i'll do it in a subsequent PR

File: lambdas/status_reports/setup.py (hunk 5 above)
Comment (sir-sigurd) [reply lambdas/status_reports/setup.py:11]: > ok, i just saw this style on some docs page and considered it "official" shrug

I think I've never seen this before :man_shrugging:.

**Comment by nl0** [comment]
> What about tests? 🙂

well...

File: lambdas/status_reports/src/t4_lambda_status_reports/__init__.py (hunk 6 above)
Comment (nl0) [reply lambdas/status_reports/src/t4_lambda_status_reports/__init__.py:268]: but these args are unused, so my local linter(ish) barfs

File: lambdas/status_reports/requirements.txt (hunk 4 above)
Comment (nl0) [reply lambdas/status_reports/requirements.txt]: > Well I think we already discussed that we use 3.7 so we can easily move code between components

i dont remember that discussion, but well, this makes *some* sense, tho idk about moving anything from lambdas to the client, for example, and getting stuck with some really old runtime is quite frustrating

**Comment by sir-sigurd** [comment]
> > What about tests? slightly_smiling_face
> 
> well...

at least you could add a stub test and add it to CI so at least we can be sure it's installable

**Comment by nl0** [comment]
> > > What about tests? slightly_smiling_face
> > 
> > 
> > well...
> 
> at least you could add a stub test and add it to CI so at least we can be sure it's installable

ok i'll add a stub

**Review by sir-sigurd — APPROVED** [review]

**Pull request merged**